frontend/catalyst/debug/printing.py

```diff
@@ -15,40 +15,77 @@
 """This module enables runtime printing of program values."""
 
 import builtins
+from functools import partial
 
 import jax
 
+from catalyst.api_extensions.callbacks import debug_callback
 from catalyst.jax_primitives import print_p
 from catalyst.tracing.contexts import EvaluationContext
 
 
 # pylint: disable=redefined-builtin
-def print(x, memref=False):
+def print(fmt, *args, **kwargs):
+    """A print function.
+
+    This print function works with python callbacks, so all arguments will be printed in the same
+    way as the :func:`builtins.print` function prints them.
+
+    One difference is that if the first argument is a string, it may be used similarly to an
+    fstring. Like so:
+
+    .. code-block:: python
+
+        @qjit
+        def cir(a, b, c):
+            debug.print("c={c} b={b} a={a}", a=a, b=b, c=c)
+
+        cir(1, 2, 3)
+    >>> c=3 b=2 a=1
+    """
+
+    if isinstance(fmt, str):
+        debug_callback(partial(_format_print_callback, fmt))(*args, **kwargs)
+        return
+
+    debug_callback(_print_callback)(fmt, *args, **kwargs)
+
+
+def _format_print_callback(fmt: str, *args, **kwargs):
+    """
+    This function has been modified from its original form in the JAX project at
+    https://github.com/google/jax/blob/5eeebf2c1829bf3c66c947b6e12464a779434e29/jax/_src/debugging.py#L269
+    version released under the Apache License, Version 2.0, with the following copyright notice:
+    Copyright 2022 The Jax Authors.
+    """
+    builtins.print(fmt.format(*args, **kwargs))
+
+
+def _print_callback(*args, **kwargs):
+    """Print without formatting"""
+    builtins.print(*args, **kwargs)
+
+
+# pylint: disable=redefined-builtin
+def print_memref(x):
     """A :func:`qjit` compatible print function for printing values at runtime.
 
-    Enables printing of numeric values at runtime. Can also print objects or strings as constants.
+    Enables printing of numeric values at runtime.
 
     Args:
-        x (jax.Array, Any): A single jax array whose numeric values are printed at runtime, or any
-            object whose string representation will be treated as a constant and printed at runtime.
-        memref (Optional[bool]): When set to ``True``, additional information about how the array is
```

Comment:
We should preserve the description of this, even if the flag is removed since that is now the behaviour of the function.

Reply:
I didn't remove it, it is on `print_memref` below.

Reply:
Ah, I am not entirely sure what you mean here. You just want everything after `memref (Optional[bool])` to be included in the docstring?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/186ab31134a57e8d86cd4f0296fea4286c3d65b6

Reply:
Yep that's what I meant, although it shouldn't be listed under the `x` arg, just in the general docstring :)

Reply:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/fb8132728b07accacbbb1dde1e5bd5b832d03afe